### PR TITLE
Wallet resync does nothing

### PR DESCRIPTION
**Pull request by banklesss**
When the resync button of settings is clicked, the wallet selection screen is displayed. Because the wallet is deselected when the wallet selection screen is displayed, no action was being applied.

Now, when the resync button is clicked, the resync action is applied and the tx list is displayed.

### CODE REVIEW

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx [hunk 1]
```diff
@@ -89,7 +85,7 @@ export const WalletSettingsScreen = () => {
 
       <SettingsSection>
         <NavigatedSettingsItem label={strings.removeWallet} navigateTo="remove-wallet" />
-        <ResyncButton />
+        <ResyncButton wallet={wallet} />
```
Comment (stackchain) [inline src/Settings/WalletSettings/WalletSettingsScreen.tsx]: imo -> move the confirmation dialog to `onClick` and if the user confirms, then fire the hook, is it a button, or is it a menu item? Anyway, I'd expect to have `onClick` API here.

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx [hunk 2]
```diff
@@ -208,71 +204,16 @@ const getWalletType = (implementationId: WalletImplementationId): MessageDescrip
...
-          mutation.mutate()
-        })
-      }
-    },
-    ...mutation,
-  }
-}
-
-const ResyncButton = () => {
+const ResyncButton = ({wallet}: {wallet: YoroiWallet}) => {
```
Comment (stackchain) [inline src/Settings/WalletSettings/WalletSettingsScreen.tsx]: onclick only

File: src/hooks/index.ts [hunk 3]
```diff
@@ -971,3 +977,58 @@ export const useBalances = (wallet: YoroiWallet): YoroiAmounts => {
...
+export const useResync = (wallet, options?: UseMutationOptions<void, Error>) => {
+  const intl = useIntl()
+  const {resetToTxHistory} = useWalletNavigation()
+  const mutation = useMutation({
+    mutationFn: () => wallet.resync(),
+    ...options,
+  })
+
+  return {
+    resyncWithConfirmation: async () => {
```
Comment (stackchain) [inline src/hooks/index.ts]: the hook should not expect user interaction

File: src/hooks/index.ts [hunk 4]
```diff
@@ -971,3 +977,58 @@ export const useBalances = (wallet: YoroiWallet): YoroiAmounts => {
...
+        InteractionManager.runAfterInteractions(() => {
+          mutation.mutate()
+        })
+      }
+    },
+    ...mutation,
+  }
+}
+
+export const useLogout = (options?: UseMutationOptions<void, Error>) => {
```
Comment (stackchain) [inline src/hooks/index.ts]: do we need to move the useLogout in this PR ?

**Review by stackchain — CHANGES_REQUESTED** [review]

**Comment by banklesss** [comment]
Draft until this is merged: https://github.com/Emurgo/yoroi-mobile/pull/2148

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx [hunk 5]
```diff
@@ -27,11 +28,14 @@ import {
 export const WalletSettingsScreen = () => {
   const intl = useIntl()
   const strings = useStrings()
-  const {navigation, resetToWalletSelection} = useWalletNavigation()
+  const {navigation, resetToWalletSelection, resetToTxHistory} = useWalletNavigation()
   const isSystemAuthEnabled = useSelector(isSystemAuthEnabledSelector)
   const wallet = useSelectedWallet()
   const walletName = useWalletName(wallet)
   const easyConfirmationEnabled = useEasyConfirmationEnabled(wallet)
+  const {resync, isLoading: isResyncLoading} = useResync(wallet, {
```
Comment (wolverineks) [inline src/Settings/WalletSettings/WalletSettingsScreen.tsx]: move all this stuff into the button

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx [hunk 6]
```diff
@@ -89,7 +100,7 @@ export const WalletSettingsScreen = () => {
 
       <SettingsSection>
         <NavigatedSettingsItem label={strings.removeWallet} navigateTo="remove-wallet" />
-        <ResyncButton />
+        <ResyncButton onClick={onResync} isLoading={isResyncLoading} />
```
Comment (wolverineks) [inline src/Settings/WalletSettings/WalletSettingsScreen.tsx]: onPress. try to replicate the API of the underlying component

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx [hunk 7]
```diff
@@ -100,10 +111,86 @@ export const WalletSettingsScreen = () => {
...
+  return <PressableSettingsItem label={strings.logout} onPress={logoutWithConfirmation} disabled={isLoading} />
+}
+
+const ResyncButton = ({onClick, isLoading}: {onClick: () => void; isLoading: boolean}) => {
+  const strings = useStrings()
+
+  return <PressableSettingsItem label={strings.resync} onPress={onClick} disabled={isLoading} />
+}
+
+const LoadingOverlay = ({loading}: {loading: boolean}) => {
```
Comment (wolverineks) [inline src/Settings/WalletSettings/WalletSettingsScreen.tsx]: why was this added?

Comment (banklesss) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: Now the resync action is called when the user is still in settings and when the resync is done the user goes back to tx list.

Comment (banklesss) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: That is the reason the loading variable is used in that way and consequently the hook is outside of the button component.

Comment (wolverineks) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: why not navigate immediately?

Comment (banklesss) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: utxos and txs are cleared on resync and after tx history is displayed they are synced again. Doing it at the same time resulted in the txs list not loading smoothly.

Comment (wolverineks) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: why doesnt it load smoothly?

**Comment by stackchain** [comment]
tested the 2 versions, and both look fine for me (resync+redirect/redirect+resync) - The difference is the balance and the locked deposit, on the 2nd version (redirecting right after calling the resync) the amount goes to 0 and then back to the value.

File: src/Settings/WalletSettings/WalletSettingsScreen.tsx (hunk 7 above)
Comment (banklesss) [reply src/Settings/WalletSettings/WalletSettingsScreen.tsx]: I committed a refactor where the immediate navigation works smoothly, so we can get the best option

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 8]
```diff
@@ -305,6 +305,12 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
     this.save()
   }
 
+  async resync() {
+    await this.clear()
+    super.resync()
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:313]: is it async?

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 9]
```diff
@@ -305,6 +305,12 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
...
+    this.save()
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts]: return / await?

**Comment by wolverineks** [comment]
~`Wallet#resync`~ `TransactionCache#resetState` doesnt actually do anything

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 8 above)
Comment (banklesss) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:313]: hi @stackchain, `super.resync()` calls `transactionCache.updateState()` and it updates `transactionCache` `state` variable. So its a sync operation.

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 9 above)
Comment (banklesss) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts]: await added, save return a promise

**Comment by banklesss** [comment]
applied this https://github.com/Emurgo/yoroi-mobile/pull/2198#discussion_r1015412819 @wolverineks

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 10]
```diff
@@ -832,7 +832,7 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
     return utxos.map(toUTXO)
   }
 
-  async clearUTXOs() {
+  async clearUtxoState() {
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts]: delete this

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 11]
```diff
@@ -97,7 +97,7 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
 
   async clear() {
     await this.transactionCache?.clear()
-    await this.clearUTXOs()
+    await this.clearUtxoState()
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts]: `this.utxoStorage.clearUtxoState()`

**Review by stackchain — APPROVED** [review]
